### PR TITLE
bootstrap: rename 'user' profile to 'dist'

### LINKED ISSUE #112074: bootstrap: Rename `profile = user` to `profile = dist`
`config.user.toml` currently says: https://github.com/rust-lang/rust/blob/20ca24e3c59dff52f76f1566c121c134e3a9b287/src/bootstrap/defaults/config.user.toml#L1
"users" is so ambiguous as to be meaningless. What I originally intended was something like "people building from source once and installing it locally", like Gentoo users or something. I think "dist" is a much better name for this profile: it makes it more clear why CI uses the profile and it hopefully will make it easier for distros to find and understand.

cc @cuviper - you said in https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Redesign.20bootstrap.20stages.20compiler-team.23619/near/351408631 this would be helpful, do you still think that? The only other name I can think of is something like `install` but I'm not sure that's a good description of what CI is doing with the profile.

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #112074

Unfortunately a big chunk of the diff is adding `PartialEq/Eq/Debug` impls so we can `assert_eq` but I think better to have them in the long run.
For back compat, ensure `"maintainer"`, `"user"` and `"dist"` are all parsed as `Profile::Dist`.

r? @jyn514 
cc @AnakinSkywalkeer who worked on the previous attempt at this

### CODE REVIEW

File: src/bootstrap/config/tests.rs [hunk 1]
```diff
@@ -159,3 +159,8 @@ fn override_toml_duplicate() {
         |&_| toml::from_str("changelog-seen = 0").unwrap(),
     );
 }
+
+#[test]
+fn profile_user_dist() {
+    assert_eq!(parse("profile = \"user\""), parse("profile = \"dist\""));
```
Comment (jyn514) [inline src/bootstrap/config/tests.rs:165]: this ends up calling `Config::parse_inner`, which in tests never reads anything from disk, and so the test is useless. it also doesn't test the python code.
```
; cat config.toml 
profile = "user"
changelog-seen = 2
; x c            
Traceback (most recent call last):
  File "/Users/jyn/src/rust/x.py", line 50, in <module>
    bootstrap.main()
  File "/Users/jyn/src/rust/src/bootstrap/bootstrap.py", line 1107, in main
    bootstrap(args)
  File "/Users/jyn/src/rust/src/bootstrap/bootstrap.py", line 1060, in bootstrap
    with open(include_path) as included_toml:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/jyn/src/rust/src/bootstrap/defaults/config.user.toml'
```
the right place to modify is `parse_inner`, like i said in the issue
https://github.com/rust-lang/rust/blob/4d74d73569e07672b101c248a2dcc396230a3da4/src/bootstrap/config.rs#L1137

File: src/bootstrap/config.rs [hunk 2]
```diff
@@ -50,7 +50,7 @@ pub enum DryRun {
     UserSelected,
 }
 
-#[derive(Copy, Clone, Default)]
+#[derive(Copy, Clone, Default, Debug, PartialEq, Eq)]
```
Comment (jyn514) [inline src/bootstrap/config.rs:53]: i am not sure how much these derives actually buy us. i don't think comparing the whole Config is particularly useful (see below for how the test is broken today) and it means that the compile times go up.

File: src/bootstrap/config/tests.rs (hunk 1 above)
Comment (clubby789) [reply src/bootstrap/config/tests.rs]: Modified `parse_inner` and updated the test, verifying it works. I'm not sure how to test the Python side with the way I've introduced the aliases

File: src/bootstrap/config.rs [hunk 3]
```diff
@@ -1129,6 +1130,17 @@ impl Config {
...
+            // profiles to be renamed while maintaining back compatibility
+            // Keep in sync with `profile_aliases` in bootstrap.py
+            static PROFILE_ALIASES: OnceLock<HashMap<&str, &str>> = OnceLock::new();
+            let include = match PROFILE_ALIASES
+                .get_or_init(|| [("user", "dist")].into_iter().collect())
+                .get(include.as_str())
+            {
+                Some(alias) => alias,
+                None => include.as_str(),
+            };
```
Comment (jyn514) [inline src/bootstrap/config.rs:1143]: i wouldn't bother with the static, bootstrap isn't perf-sensitive.
```suggestion
            let profile_aliases = HashMap::from([("user", "dist")]);
            let include = profile_aliases.get(include.as_str()).unwrap_or(include.as_str())
```

**Review by jyn514 — APPROVED** [review]
thanks, this looks good! r=me with or without the nit fixed :)

**Pull request merged**